src/engraving/libmscore/measure.cpp

```diff
@@ -4043,12 +4037,22 @@ float Measure::durationStretch(Fraction curTicks, const Fraction minTicks) const
         }
     }
 
-    if (minTicks < shortNoteThreshold) {
-        // Reduces the slope of the spacing curve in case very short notes are present.
-        // Avoids having the longer notes too wide.
-        qreal reduction = qMax((1 - 0.2 * log2(qreal(shortNoteThreshold.ticks()) / qreal(minTicks.ticks()))), 0.3);
-        // The numbers (and the formula itself) are purely empirical.
-        slope = slope * reduction;
+    //------------------------------------------------------------------------
+    // Prevent long notes from exploding in the presence of very short ones.
+    // maxRatio defines the maximum accepted ratio between the longest and
+    // shortest note of the system. If this ratio is exceeded, the subsequent
+    // calculations ensure that the highest ratio is renormalized to maxRatio
+    // and all other ratios are scaled accordingly.
+    //------------------------------------------------------------------------
+    static constexpr double maxRatio = 32.0;
+    double minSysTicks = double(minTicks.ticks());
+    double maxSysTicks = double(system()->maxSysTicks().ticks());
+    double maxSysRatio = maxSysTicks / minSysTicks;
+    double ratio = double(curTicks.ticks()) / minSysTicks;
+    if (maxSysRatio > maxRatio) {
+        double A = minSysTicks * (maxRatio - 1) / (maxSysTicks - minSysTicks);
+        double B = (maxSysTicks - maxRatio * minSysTicks) / (maxSysTicks - minSysTicks);
```

Comment:
for these three lines, it's better to use parentheses to make sure the order of operations is what you want it to be. There have been a number of bugs I've been fixing recently that stem from people being unclear about the order of operations, so it's better to write
`((maxSysTicks - maxRatio) * minSysTicks) / (maxSysTicks - minSysTicks);`, for example